### PR TITLE
Fix issue #2441 & #2442 

### PR DESCRIPTION
**Pull request by DamaniN**
##### Pull Request Details:

* Type: **Enhancement**

* Impact: **High** 

* Reference to related issue (URL):

#2441 

* How was this pull request tested?

Restored system from replica cluster running Rubrik CDM v5.2

* Brief description of the changes in this pull request:

The RBS agent file name changed in Rubrik CDM v5.1. Failed over to the new file name if the old file doesn't work.

### CODE REVIEW

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh [hunk 1]
```diff
@@ -25,12 +25,18 @@ while true; do
...
+CDM_TAR_FILE=$CDM_SUNOS_TAR
 cd /tmp
-/usr/bin/curl $v -kLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_SUNOS_TAR} 
-StopIfError "Could not download https://${CDM_CLUSTER_IP}/connector/${CDM_SUNOS_TAR}"
+/usr/bin/curl $v -fskLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_TAR_FILE} 
+if [[ $? -gt 0 ]];  then
+    CDM_TAR_FILE=$CDM_SOLARIS_TAR
+    /usr/bin/curl $v -fkLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_TAR_FILE} 
+fi
+StopIfError "Could not download cluster certificate extraction."
```
Comment (jsmeix) [inline usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh:36]: I think in general you could make error messages more informative for the user
when you provide some variable values that tell the user what
in this case here that `cluster certificate extraction` thingy actually is e.g. like
```
StopIfError "Could not download cluster certificate extraction https://${CDM_CLUSTER_IP}/connector/${CDM_TAR_FILE}"
```

**Comment by jsmeix** [comment]
@rear/contributors 
I would appreciate a second review from any other ReaR contributor.

**Comment by DamaniN** [comment]
> It seems via this pull request the file
> usr/share/rear/verify/CDM/default/400_verify_cdm.sh
> gets deleted.
> Is that intentional?

Yes this is intentional. That file was causing the Rubrik agent to be started twice. The functionality was moved to another file in an earlier update but the file was never removed at that time. Starting the agent twice was contributing to #2441.

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh (hunk 1 above)
Comment (DamaniN) [reply usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh]: Updated error message in 7b1e26b9b8c43934e46d8a1f95c8b85168a807df.

**Comment by DamaniN** [comment]
> I do not use `BACKUP=CDM` so I cannot test anything here but
> from plain looking at the code things look OK to me.
> 
> I wished there was a comment in the code
> (cf. "Code must be easy to understand" in https://github.com/rear/rear/wiki/Coding-Style)
> that at least mentiones the issue #2441
> or even better directly tells about the source of this problem
> why two file names must be tried to dowload
> for example something like
> 
> ```
> # The name of the tar file that is being downloaded has changed in Rubrik CDM v5.1.
> # Before Rubrik CDM v5.1 it was rubrik-agent-sunos5.10.sparc.tar.gz
> # since Rubrik CDM v5.1 it is rubrik-agent-solaris.sparc.tar.gz
> # cf. https://github.com/rear/rear/issues/2441
> ```

Added this note in commit 7b1e26b9b8c43934e46d8a1f95c8b85168a807df.

**Comment by DamaniN** [comment]
d86ca29cd561d49bb00ee64ee120a1ff042ae18a also fixes #2442.

**Comment by jsmeix** [comment]
@DamaniN 
thank you for your prompt replies and your further code enhancements.

Now things are clear to me so I approve it.

**Comment by jsmeix** [comment]
@rear/contributors 
a second review from another ReaR contributor would be nice to have.
If there is none and if there are no objections I would "just merge" it tomorrow afternoon.

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh [hunk 2]
```diff
@@ -24,13 +24,24 @@ while true; do
...
 
+# The name of the tar file that is being downloaded has changed in Rubrik CDM v5.1.
+# Before Rubrik CDM v5.1 it was rubrik-agent-sunos5.10.sparc.tar.gz
+# since Rubrik CDM v5.1 it is rubrik-agent-solaris.sparc.tar.gz
+# cf. https://github.com/rear/rear/issues/2441
+
 CDM_SUNOS_TAR=rubrik-agent-sunos5.10.sparc.tar.gz
+CDM_SOLARIS_TAR=rubrik-agent-solaris.sparc.tar.gz
+CDM_TAR_FILE=$CDM_SUNOS_TAR
 cd /tmp
```
Comment (schlomo) [inline usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh:35]: Please use `$TMP_DIR` for temporary files

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh [hunk 3]
```diff
@@ -24,13 +24,24 @@ while true; do
...
-/usr/bin/curl $v -kLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_SUNOS_TAR} 
-StopIfError "Could not download https://${CDM_CLUSTER_IP}/connector/${CDM_SUNOS_TAR}"
+/usr/bin/curl $v -fskLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_TAR_FILE}
```
Comment (schlomo) [inline usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh:36]: `/usr/bin/` not required

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh [hunk 4]
```diff
@@ -24,13 +24,24 @@ while true; do
...
+CDM_TAR_FILE=$CDM_SUNOS_TAR
 cd /tmp
-/usr/bin/curl $v -kLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_SUNOS_TAR} 
-StopIfError "Could not download https://${CDM_CLUSTER_IP}/connector/${CDM_SUNOS_TAR}"
+/usr/bin/curl $v -fskLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_TAR_FILE} 
+if [[ $? -gt 0 ]];  then
+    CDM_TAR_FILE=$CDM_SOLARIS_TAR
+    /usr/bin/curl $v -fkLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_TAR_FILE} 
+fi
+StopIfError "Could not download Rubrik agent from https://${CDM_CLUSTER_IP}/connector/${CDM_SUNOS_TAR} or https://${CDM_CLUSTER_IP}/connector/${CDM_SOLARIS_TAR}."
```
Comment (schlomo) [inline usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh:41]: Alternative approach could be to simply give `curl` both URLs and check that you got at least one file. That will make it easier to adapt the code to a future version of CDM that will call the Linux client `rubrik-agent-linux.tar.gz` 😄

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh [hunk 5]
```diff
@@ -24,13 +24,24 @@ while true; do
...
 
-/usr/bin/tar $v -xzf  $CDM_SUNOS_TAR
-StopIfError "Could not extract $CDM_SUNOS_TAR"
+/usr/bin/tar $v -xzf  $CDM_TAR_FILE
```
Comment (schlomo) [inline usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh:43]: `tar` is also in `$PATH`

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh [hunk 6]
```diff
@@ -39,5 +39,8 @@ StopIfError "Could not copy replica CDM cluster certificate"
 
 /usr/bin/chmod $v 600 ${CDM_KEYS_DIR}/rubrik.crt
 
+mv $v ${CDM_KEYS_DIR}/agent.crt ${CDM_KEYS_DIR}/agent.crt.orig
+mv $v ${CDM_KEYS_DIR}/agent.pem ${CDM_KEYS_DIR}/agent.pem.orig
+/etc/rubrik/rba-keygen.sh
```
Comment (gdha) [inline usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh:55]: @DamaniN  Why is this required?

**Comment by jsmeix** [comment]
@gdha 
for the reason of the removal of usr/share/rear/verify/CDM/default/400_verify_cdm.sh
see https://github.com/rear/rear/pull/2445#issuecomment-651883744

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh (hunk 2 above)
Comment (jsmeix) [reply usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh:41]: In general `cd /some/dir` in a ReaR script is problematic.
Usually it should be `pushd /some/dir` with matching `popd`.

Regarding hardcoded `/tmp` I did
https://github.com/rear/rear/issues/2460

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh [hunk 7]
```diff
@@ -24,13 +24,24 @@ while true; do
...
 CDM_SUNOS_TAR=rubrik-agent-sunos5.10.sparc.tar.gz
+CDM_SOLARIS_TAR=rubrik-agent-solaris.sparc.tar.gz
+CDM_TAR_FILE=$CDM_SUNOS_TAR
 cd /tmp
-/usr/bin/curl $v -kLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_SUNOS_TAR} 
-StopIfError "Could not download https://${CDM_CLUSTER_IP}/connector/${CDM_SUNOS_TAR}"
+/usr/bin/curl $v -fskLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_TAR_FILE} 
+if [[ $? -gt 0 ]];  then
+    CDM_TAR_FILE=$CDM_SOLARIS_TAR
+    /usr/bin/curl $v -fkLOJ https://${CDM_CLUSTER_IP}/connector/${CDM_TAR_FILE}
```
[the remark this answers is not in the document]
Comment (jsmeix) [reply usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh]: fixed

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh (hunk 3 above)
Comment (jsmeix) [reply usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh]: fixed

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh (hunk 5 above)
Comment (jsmeix) [reply usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh]: fixed

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh (hunk 4 above)
Comment (jsmeix) [reply usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh]: I will implement something like that tomorrow...

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh (hunk 2 above)
Comment (jsmeix) [reply usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh:41]: At other places we have code like
```
  ( cd /some/dir
    some commands
  )
```
which uses a subshell
but I think normally code like
```
    pushd /seome/dir
    some commands
    popd
```
looks somehow "better" to me.

I wonder if we should perhaps do in the `Source` function
something like
```
function Source () {
    local current_dir=$( pwd )
    ...
    source "$source_file"
    source_return_code=$?
    cd $current_dir
```
to ensure that after each script
we are back in ReaR's usual working directory?

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh (hunk 4 above)
Comment (jsmeix) [reply usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh]: Puh!
That became almost a complete rewrite.

**Comment by jsmeix** [comment]
I tested it as far as I could without a Rubrik (CDM) cluster environment
by simulating things as far as possible with reasonable effort.
To me things look OK now.

If there are no objections I would like to merge it as is tomorrow afternoon.
We are in early ReaR 2.7 development phase so we can fix things if needed.

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh (hunk 6 above)
Comment (jsmeix) [reply usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh]: I added a LogPrintError message so theer is at least some user info.
That is all I could do because I do not have a Rubrik (CDM) cluster environment.

File: usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh (hunk 2 above)
Comment (jsmeix) [reply usr/share/rear/verify/CDM/default/410_use_replica_cdm_cluster_cert.sh]: I did https://github.com/rear/rear/issues/2461
